src/rp2_common/pico_rand/rand.c

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) 2022 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/*  xoroshiro128ss(), rotl():
+
+    Written in 2018 by David Blackman and Sebastiano Vigna (vigna@acm.org)
+
+    To the extent possible under law, the author has dedicated all copyright
+    and related and neighboring rights to this software to the public domain
+    worldwide. This software is distributed without any warranty.
+
+    See <http://creativecommons.org/publicdomain/zero/1.0/>
+
+    splitmix64() implementation:
+
+    Written in 2015 by Sebastiano Vigna (vigna@acm.org)
+    To the extent possible under law, the author has dedicated all copyright
+    and related and neighboring rights to this software to the public domain
+    worldwide. This software is distributed without any warranty.
+
+    See <http://creativecommons.org/publicdomain/zero/1.0/>
+*/
+
+#include "pico/rand.h"
+#include "pico/unique_id.h"
+#include "hardware/timer.h"
+#include "hardware/clocks.h"
+#include "hardware/structs/rosc.h"
+#include "hardware/sync.h"
+
+// Hashing the selected area of RAM is done the first time
+// a random number is requested. If there is any chance the first RN will be
+// requested from interrupt context only a small amount of RAM should be hashed
+// hence the 1k default here which will take ~300us on RP2040 at 125Mhz
+
+// PICO_CONFIG: PICO_RAND_DISABLE_ROSC_CHECKS, Dsiable assertions that the ROSC is running when generating random numbers which will drastically reduce randomness, type=bool, default=0, group=pico_rand
+
+// PICO_CONFIG: PICO_RAND_RAM_HASH_END, end of address in RAM rand (non-inclusive) to hash during pico_rand initialization, default=SRAM_END, group=pico_rand
+#ifndef PICO_RAND_RAM_HASH_END
+#define PICO_RAND_RAM_HASH_END     SRAM_END
+#endif
+// PICO_CONFIG: PICO_RAND_RAM_HASH_START, end of address in RAM rand (non-inclusive) to hash during pico_rand initialization, default=SRAM_END, group=pico_rand
```

Comment:
"end of address in RAM rand" -> "start of address in RAM rand" ? Also, the `default=` seems incorrect?